src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Setup/ConfiguredSetupComponentSnapshot.cs

```diff
@@ -0,0 +1,155 @@
+
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Setup;
+
+/// <summary>
+/// Immutable snapshot of the components required by a configured project.
+/// </summary>
+internal sealed class ConfiguredSetupComponentSnapshot
+{
+    private const string WebComponentId = "Microsoft.VisualStudio.Component.Web";
+
+    /// <summary>
+    /// Maps .NET Core <c>TargetFrameworkVersion</c> to the corresponding VS Setup component ID for that version's runtime.
+    /// </summary>
+    public static readonly ImmutableDictionary<string, string> ComponentIdByRuntimeVersion = ImmutableStringDictionary<string>.EmptyOrdinalIgnoreCase
+        .Add("v2.0", "Microsoft.Net.Core.Component.SDK.2.1")
+        .Add("v2.1", "Microsoft.Net.Core.Component.SDK.2.1")
+        .Add("v2.2", "Microsoft.Net.Core.Component.SDK.2.1")
+        .Add("v3.0", "Microsoft.NetCore.Component.Runtime.3.1")
+        .Add("v3.1", "Microsoft.NetCore.Component.Runtime.3.1")
+        .Add("v5.0", "Microsoft.NetCore.Component.Runtime.5.0")
+        .Add("v6.0", "Microsoft.NetCore.Component.Runtime.6.0")
+        .Add("v7.0", "Microsoft.NetCore.Component.Runtime.7.0")
+        .Add("v8.0", "Microsoft.NetCore.Component.Runtime.8.0");
+
+    public static ConfiguredSetupComponentSnapshot Empty { get; } = new();
+
+    private readonly bool _requiresWebComponent;
+    private readonly string? _netCoreTargetFrameworkVersion;
+    private readonly ImmutableHashSet<string> _suggestedWorkloadComponentIds;
+
+    public ImmutableHashSet<string> ComponentIds { get; }
+    
+    public bool IsEmpty { get; }
+
+    private ConfiguredSetupComponentSnapshot()
+    {
+        _suggestedWorkloadComponentIds = ComponentIds = ImmutableStringHashSet.EmptyVisualStudioSetupComponentIds;
+        IsEmpty = true;
+    }
+    
+    private ConfiguredSetupComponentSnapshot(bool requiresWebComponent, string? netCoreTargetFrameworkVersion, ImmutableHashSet<string> suggestedWorkloadComponentIds)
+    {
+        _requiresWebComponent = requiresWebComponent;
+        _netCoreTargetFrameworkVersion = netCoreTargetFrameworkVersion;
+        _suggestedWorkloadComponentIds = suggestedWorkloadComponentIds;
+
+        ImmutableHashSet<string> componentIds = suggestedWorkloadComponentIds;
+
+        if (requiresWebComponent)
+        {
+            componentIds = componentIds.Add(WebComponentId);
+        }
+
+        if (netCoreTargetFrameworkVersion is not null && ComponentIdByRuntimeVersion.TryGetValue(netCoreTargetFrameworkVersion, out string? runtimeComponentId))
+        {
+            componentIds = componentIds.Add(runtimeComponentId);
+        }
+
+        ComponentIds = componentIds;
+    }
+
+    public ConfiguredSetupComponentSnapshot Update(
+        IProjectSubscriptionUpdate evaluationUpdate,
+        IProjectSubscriptionUpdate buildUpdate,
+        IProjectCapabilitiesSnapshot capabilities)
+    {
+        if (IsEmpty |
```

Comment:
Would prefer logical or over bitwise or here.

Reply:
How would you write it?

Reply:
How about:

```suggestion
        // We use bitwise | here instead of logical || to prevent short circuiting.
        if (IsEmpty |
```

Reply:
Comment is also good. Sadly the way to write with logical operations involves pulling all conditions out in variable to make sure they are executed before "if". Compiler is just too smart.